dotnet/runtime
# [browser] Migrate more Blazor features, prepare JavaScript API for Blazor cleanup

Goal of this PR is to prepare everything needed to clean-up startup on Blazor side to the final shape.
Changes needs to be backward compatible with Blazor.

- `GlobalizationMode` & `ICUDataMode` merged
  - Resulting type is string based enum
  - Contains separate values for different ICU modes ("all", "sharded", "custom" instead of "icu")
  - Value "auto" has been removed
- Integrate library initializers
  - Allows for nuget packages to hook into application startup
  - Nuget package can contains files in name pattern `*.lib.module.js` in wwwroot
  - Such es6 modules are imported during startup
  - They can export various callbacks
    - On runtime side we call `onRuntimeConfigLoaded(config: MonoConfig)` and `onRuntimeReady(api: RuntimeAPI)`
    - On blazor side, `beforeStart` and `afterStarted` are called
  - Library initializers can be downloaded either right after config is loaded or before runtime ready promise resolved. By default boot json generator checks for content to see if library initializer must be downloaded earlier. This behavior can be overriden by putting library initializer in `@(WasmLibraryInitializerOnRuntimeConfigLoaded)` or `@(WasmLibraryInitializerOnRuntimeReady)`
  - Library initializers can invoked using `invokeLibraryInitializers`
- New set of Wasm.Build.Tests using prepared wasm app
  - The purpose is to have a coded application and test behavior/interaction with it using playwright
  - Test lazy assembly loading
  - Test loading satellite assembly
  - Test library initializer
  - Test appsettings based on app environment
- Run all Wasm.Build.Tests in specified namespace using `XunitNamespace` msbuild property
- Set debug level in Wasm SDK apps
  - Use `WasmDebugLevel` to specify debug level (the same as `WasmApp.targets`)
  - Default values (when `WasmDebugLevel` is not set) for .NET >= 8.0
    - Build   (debug) => debugBuild=true & debugLevel=-1 => `-1`
    - Build   (release) => debugBuild=true & debugLevel=0  => `0`
    - Publish (debug) => debugBuild=false & debugLevel=-1 => `0`
    - Publish (release) => debugBuild=false & debugLevel=0  => `0`
- Fix resolving URL (config, library initializer) when baseURI contains query string



Finalize
- [x] Run Blazor E2E to verify no breaking changes

Contributes to https://github.com/dotnet/runtime/issues/70762
Contributes to https://github.com/dotnet/runtime/issues/87178
